controllers/etcdcluster_controller.go

```diff
@@ -1391,39 +1422,70 @@ func expectedURLForPeer(cluster *etcdv1alpha1.EtcdCluster, peerName string) stri
 	)
 }
 
-func (r *EtcdClusterReconciler) defragCronJob(log logr.Logger, cluster *etcdv1alpha1.EtcdCluster, members *[]etcd.Member) {
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second*25)
+func (r *EtcdClusterReconciler) defragWithThresholdCronJob(log logr.Logger, cluster *etcdv1alpha1.EtcdCluster) {
+	ctx, cancel, etcdClient, members, ok := r.setupDefragDeps(log, cluster)
```

Comment:
I think it's a bit cleaner to create both the required context's outside of `setupDefragDeps`, so it's really clear what context is being used where and what timeouts are set.

It would also be nice to add a comment on why we need a new context (i.e "we can't use `reconciles` ctx, as it will have likely expired when this cronjob is ran").

Reply:
I agree that it was a bit strange to return the ctx, cancel from the method (also I think there was a bug with defer cancel() as well). I've moved the creation of the defrag context to each method https://github.com/storageos/etcd-cluster-operator/pull/37/files#diff-4d256d176300fb45905cd320c72758685ad4f268407980cfd0564bb59504bd26R1444 I did not move the tls context out though, imo it belongs in the method as a implementation detail. But happy to move it outside and pass it in if you think it's clearer.